src/express/commands/milestone-base.js

```diff
@@ -0,0 +1,341 @@
+/* eslint-disable dot-notation */
+import { loadDevnetConfig, splitToArray } from '../common/config-utils'
+import { timer } from '../common/time-utils'
+import { maxRetries, runCommand } from '../common/remote-worker'
+import {
+  getBlock,
+  getPeerLength,
+  createClusters,
+  getEnode,
+  fetchLatestMilestone,
+  joinAllPeers,
+  validateProposer,
+  validateFinalizedBlock,
+  checkForRewind
+} from '../common/milestone-utils'
+
+const milestoneLength = 64
+const queryTimer = (milestoneLength / 4) * 1000
+
+export async function milestoneBase() {
+  // NOTE: Make sure bor branch has logic for hardcoded primary validator
+
+  require('dotenv').config({ path: `${process.cwd()}/.env` })
+  const devnetType =
+    process.env.TF_VAR_DOCKERIZED === 'yes' ? 'docker' : 'remote'
+
+  const doc = await loadDevnetConfig(devnetType)
+
+  const borUsers = splitToArray(doc['devnetBorUsers'].toString())
+  const borHosts = splitToArray(doc['devnetBorHosts'].toString())
+
+  // Check for number of validators
+  if (doc['numOfValidators'].length < 4) {
+    console.log('📍Cannot run milestone tests on less than 4 validator nodes')
+    process.exit(1)
+  }
+
+  // Grab the enode of all the nodes
+  let enodes = []
+  let tasks = []
+  const ips = []
+  for (let i = 0; i < borUsers.length; i++) {
+    const ip = `${borUsers[i]}@${borHosts[i]}`
+    ips.push(ip)
+    tasks.push(getEnode(borUsers[i], borHosts[i]))
+  }
+
+  await Promise.all(tasks).then((values) => {
+    enodes = values
+  })
+
+  if (enodes.includes('')) {
+    console.log('📍Unable to fetch enode, exiting')
+    return
+  }
+
+  console.log('📍Rejoining clusters before performing tests')
+  let joined = await joinAllPeers(ips, enodes)
+  if (!joined) {
+    console.log('📍Unable to join peers before starting tests, exiting')
+    return
+  }
+
+  console.log('📍Rejoined clusters')
+
+  // Wait for a milestone to get proposed for verification
+  const lastMilestone = await fetchLatestMilestone(
+    milestoneLength,
+    queryTimer,
+    borHosts[0]
+  )
+  if (!lastMilestone) {
+    console.log('📍Unable to fetch latest milestone from heimdall, exiting')
+    return
+  }
+
+  console.log('📍Waiting 10s to fetch finalized blocks...')
+  await timer(10000)
```

Comment:
The timer to query milestone is quite low for bor and hence if a milestone just got proposed, we're sure that we'll get a finalised block. This can be made configurable via some global params tho.

Reply:
If it doesn't have the milestone, the function below should give an error. We shouldn't retry IMO as something went wrong with saving milestones.